src/ggml.c

```diff
@@ -20747,14 +20890,11 @@ enum ggml_opt_result ggml_opt_resume(
         struct ggml_tensor * f) {
 
     // build forward + backward compute graphs
-    struct ggml_tensor * gfbuf = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, sizeof(struct ggml_cgraph) / ggml_type_size(GGML_TYPE_I32)+ (sizeof(struct ggml_cgraph) % ggml_type_size(GGML_TYPE_I32) ? 1 : 0));
-    struct ggml_tensor * gbbuf = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, sizeof(struct ggml_cgraph) / ggml_type_size(GGML_TYPE_I32)+ (sizeof(struct ggml_cgraph) % ggml_type_size(GGML_TYPE_I32) ? 1 : 0));
-
-    struct ggml_cgraph * gf = (struct ggml_cgraph *) gfbuf->data;
-    struct ggml_cgraph * gb = (struct ggml_cgraph *) gbbuf->data;
+    struct ggml_cgraph * gf;
+    struct ggml_cgraph * gb;
 
-    *gf = ggml_build_forward (f);
-    *gb = ggml_build_backward(ctx, gf, true);
+    gf = ggml_build_forward_ctx (ctx, f, GGML_DEFAULT_GRAPH_SIZE); // TODO: make graph size configurable
+    gb = ggml_build_backward_ctx(ctx, gf, true);
```

Comment:
The `ggml_opt` interface may also need to be changed to take a graph size.

Reply:
Ok, I will take a look tomorrow if you haven't updated it yet

Reply:
Think we can simply extend `ggml_opt_params` with a new parameter:

```c
size_t size; // max graph size (both for forward and backward graphs)
```

The backward graph should always be bigger than the forward, so the user has to determine the size needed to store it.

Regarding how the user would determine the necessary graph size (not just for `ggml_opt` but in general), I guess it will be via trial and error. It should be fine, since we don't expect graphs to change dynamically so it is going to be a one-time effort to find the max size (given that the default value is not enough). We only have to make sure to report an error that the graph size has been exceeded when this happens.

Reply:
I have been thinking about how to remove the graph sizes entirely and make this automatic, but I don't think that this can be done as long as graphs are allocated in a context, because otherwise we would just be moving the problem to the context allocation (ie. even if the user doesn't need to specify the size of the graph, they would still need to calculate the size to determine how much memory to allocate for the context). So if we wanted to do this, we would need to allocate the graphs in the heap with `malloc`/etc, and to avoid reallocations between evaluations, the graph object would need to be reused between evaluations.

We should probably do this at some point, while I agree that it is not a huge problem to find the size of a graph, it shouldn't be necessary, and generally it feels a bit archaic to work with fixed size containers.